File: src/domain/services/canvas_assembler.py
"""Domain service that assembles Bounded Context Canvases from a DomainModel.

CanvasAssembler is stateless: it reads from a DomainModel aggregate and
produces immutable BoundedContextCanvas value objects.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from src.domain.models.bounded_context_canvas import (
    BoundedContextCanvas,
    CommunicationMessage,
    DomainRole,
    StrategicClassification,
)
from src.domain.models.domain_values import SubdomainClassification

if TYPE_CHECKING:
    from src.domain.models.domain_model import DomainModel

_CLASSIFICATION_TO_ROLE: dict[SubdomainClassification, DomainRole] = {
    SubdomainClassification.CORE: DomainRole.EXECUTION,
    SubdomainClassification.SUPPORTING: DomainRole.SPECIFICATION,
    SubdomainClassification.GENERIC: DomainRole.GATEWAY,
}


class CanvasAssembler:
    """Stateless domain service that assembles BC Canvases from a DomainModel."""

    @staticmethod
    def assemble(model: DomainModel) -> tuple[BoundedContextCanvas, ...]:
        """Build one canvas per bounded context in the model."""
        canvases: list[BoundedContextCanvas] = []

        for ctx in model.bounded_contexts:
            domain_cls = ctx.classification or SubdomainClassification.GENERIC

            classification = StrategicClassification(
                domain=domain_cls,
                business_model="unclassified",
                evolution="unclassified",
            )

            role = _CLASSIFICATION_TO_ROLE.get(domain_cls, DomainRole.DRAFT)

            inbound = tuple(
                CommunicationMessage(
                    message=rel.integration_pattern,
                    message_type="Event",
                    counterpart=rel.upstream,
                )
                for rel in model.context_relationships
                if rel.downstream == ctx.name
            )
            outbound = tuple(
                CommunicationMessage(
                    message=rel.integration_pattern,
                    message_type="Event",
                    counterpart=rel.downstream,
                )
                for rel in model.context_relationships
                if rel.upstream == ctx.name
            )

            ul_terms = tuple(
                (entry.term, entry.definition)
                for entry in model.ubiquitous_language.terms
                if entry.context_name == ctx.name
            )

            decisions: list[str] = []
            for agg in model.aggregate_designs:
                if agg.context_name == ctx.name:
                    decisions.extend(agg.invariants)

            canvases.append(
                BoundedContextCanvas(
                    context_name=ctx.name,
                    purpose=ctx.responsibility,
                    classification=classification,
                    domain_roles=(role,),
                    inbound_communication=inbound,
                    outbound_communication=outbound,
                    ubiquitous_language=ul_terms,
                    business_decisions=tuple(decisions),
                    assumptions=(),
                    open_questions=(),
                )
            )

        return tuple(canvases)

    @staticmethod
    def render_markdown(canvases: tuple[BoundedContextCanvas, ...]) -> str:
        """Render canvases to markdown following ddd-crew v5 format."""
        if not canvases:
            return ""

        sections = [_render_single_canvas(canvas) for canvas in canvases]
        return "\n---\n\n".join(sections)


def _render_single_canvas(canvas: BoundedContextCanvas) -> str:
    """Render a single canvas to markdown."""
    lines: list[str] = [
        f"# Bounded Context Canvas: {canvas.context_name}",
        "",
    ]
    _add_purpose(lines, canvas)
    _add_classification(lines, canvas)
    _add_domain_roles(lines, canvas)
    _add_communication(lines, "Inbound Communication", "Sender", canvas.inbound_communication)
    _add_communication(lines, "Outbound Communication", "Receiver", canvas.outbound_communication)
    _add_ul(lines, canvas)
    _add_list_section(lines, "Business Decisions & Rules", canvas.business_decisions)
    _add_list_section(lines, "Assumptions", canvas.assumptions)
    _add_list_section(lines, "Open Questions", canvas.open_questions)
    return "\n".join(lines)


def _add_purpose(lines: list[str], canvas: BoundedContextCanvas) -> None:
    lines.extend(["## Purpose", "", canvas.purpose, ""])


def _add_classification(lines: list[str], canvas: BoundedContextCanvas) -> None:
    c = canvas.classification
    lines.extend([
        "## Strategic Classification",
        "",
        "| Aspect | Value |",
        "| --- | --- |",
        f"| Domain | {c.domain.value} |",
        f"| Business Model | {c.business_model} |",
        f"| Evolution | {c.evolution} |",
        "",
    ])


def _add_domain_roles(lines: list[str], canvas: BoundedContextCanvas) -> None:
    lines.extend(["## Domain Roles", ""])
    lines.extend(f"- [x] {role.value}" for role in canvas.domain_roles)
    lines.append("")


def _add_communication(
    lines: list[str],
    heading: str,
    third_col: str,
    messages: tuple[CommunicationMessage, ...],
) -> None:
    lines.extend([f"## {heading}", ""])
    if messages:
        lines.extend([
            f"| Message | Type | {third_col} |",
            "| --- | --- | --- |",
        ])
        lines.extend(
            f"| {m.message} | {m.message_type} | {m.counterpart} |" for m in messages
        )
    else:
        lines.append("*None*")
    lines.append("")


def _add_ul(lines: list[str], canvas: BoundedContextCanvas) -> None:
    lines.extend(["## Ubiquitous Language", ""])
    if canvas.ubiquitous_language:
        lines.extend(["| Term | Definition |", "| --- | --- |"])
        lines.extend(f"| {term} | {defn} |" for term, defn in canvas.ubiquitous_language)
    else:
        lines.append("*None*")
    lines.append("")


def _add_list_section(lines: list[str], heading: str, items: tuple[str, ...]) -> None:
    lines.extend([f"## {heading}", ""])
    if items:
        lines.extend(f"- {item}" for item in items)
    else:
        lines.append("*None*")
    lines.append("")

File: internal/fitness/domain/canvas_assembler.go
package domain

import (
	"fmt"
	"strings"

	"github.com/alty-cli/alty/internal/shared/domain/ddd"
	vo "github.com/alty-cli/alty/internal/shared/domain/valueobjects"
)

var classificationToRole = map[vo.SubdomainClassification]Role{
	vo.SubdomainCore:       RoleExecution,
	vo.SubdomainSupporting: RoleSpecification,
	vo.SubdomainGeneric:    RoleGateway,
}

// Assemble builds one canvas per bounded context in the model.
func Assemble(model *ddd.DomainModel) []BoundedContextCanvas {
	var canvases []BoundedContextCanvas

	for _, ctx := range model.BoundedContexts() {
		domainCls := vo.SubdomainGeneric
		if ctx.Classification() != nil {
			domainCls = *ctx.Classification()
		}

		classification := NewStrategicClassification(domainCls, "unclassified", "unclassified")

		role, ok := classificationToRole[domainCls]
		if !ok {
			role = RoleDraft
		}

		var inbound []CommunicationMessage
		for _, rel := range model.ContextRelationships() {
			if rel.Downstream() == ctx.Name() {
				inbound = append(inbound, NewCommunicationMessage(
					rel.IntegrationPattern(), "Event", rel.Upstream()))
			}
		}

		var outbound []CommunicationMessage
		for _, rel := range model.ContextRelationships() {
			if rel.Upstream() == ctx.Name() {
				outbound = append(outbound, NewCommunicationMessage(
					rel.IntegrationPattern(), "Event", rel.Downstream()))
			}
		}

		var ulTerms [][2]string
		for _, entry := range model.UbiquitousLanguage().Terms() {
			if entry.ContextName() == ctx.Name() {
				ulTerms = append(ulTerms, [2]string{entry.Term(), entry.Definition()})
			}
		}

		var decisions []string
		for _, agg := range model.AggregateDesigns() {
			if agg.ContextName() == ctx.Name() {
				decisions = append(decisions, agg.Invariants()...)
			}
		}

		canvases = append(canvases, NewBoundedContextCanvas(
			ctx.Name(),
			ctx.Responsibility(),
			classification,
			[]Role{role},
			inbound, outbound, ulTerms, decisions,
			nil, nil,
		))
	}

	return canvases
}

// RenderMarkdown renders canvases to markdown following ddd-crew v5 format.
func RenderMarkdown(canvases []BoundedContextCanvas) string {
	if len(canvases) == 0 {
		return ""
	}

	sections := make([]string, len(canvases))
	for i, canvas := range canvases {
		sections[i] = renderSingleCanvas(canvas)
	}
	return strings.Join(sections, "\n---\n\n")
}

func renderSingleCanvas(canvas BoundedContextCanvas) string {
	var lines []string
	lines = append(lines, fmt.Sprintf("# Bounded Context Canvas: %s", canvas.contextName), "")
	addPurpose(&lines, canvas)
	addClassification(&lines, canvas)
	addRoles(&lines, canvas)
	addCommunication(&lines, "Inbound Communication", "Sender", canvas.inboundCommunication)
	addCommunication(&lines, "Outbound Communication", "Receiver", canvas.outboundCommunication)
	addUL(&lines, canvas)
	addListSection(&lines, "Business Decisions & Rules", canvas.businessDecisions)
	addListSection(&lines, "Assumptions", canvas.assumptions)
	addListSection(&lines, "Open Questions", canvas.openQuestions)
	return strings.Join(lines, "\n")
}

func addPurpose(lines *[]string, canvas BoundedContextCanvas) {
	*lines = append(*lines, "## Purpose", "", canvas.purpose, "")
}

func addClassification(lines *[]string, canvas BoundedContextCanvas) {
	c := canvas.classification
	*lines = append(*lines,
		"## Strategic Classification", "",
		"| Aspect | Value |",
		"| --- | --- |",
		fmt.Sprintf("| Domain | %s |", string(c.domain)),
		fmt.Sprintf("| Business Model | %s |", c.businessModel),
		fmt.Sprintf("| Evolution | %s |", c.evolution),
		"",
	)
}

func addRoles(lines *[]string, canvas BoundedContextCanvas) {
	*lines = append(*lines, "## Domain Roles", "")
	for _, role := range canvas.domainRoles {
		*lines = append(*lines, fmt.Sprintf("- [x] %s", string(role)))
	}
	*lines = append(*lines, "")
}

func addCommunication(lines *[]string, heading, thirdCol string, messages []CommunicationMessage) {
	*lines = append(*lines, fmt.Sprintf("## %s", heading), "")
	if len(messages) > 0 {
		*lines = append(*lines,
			fmt.Sprintf("| Message | Type | %s |", thirdCol),
			"| --- | --- | --- |",
		)
		for _, m := range messages {
			*lines = append(*lines, fmt.Sprintf("| %s | %s | %s |", m.message, m.messageType, m.counterpart))
		}
	} else {
		*lines = append(*lines, "*None*")
	}
	*lines = append(*lines, "")
}

func addUL(lines *[]string, canvas BoundedContextCanvas) {
	*lines = append(*lines, "## Ubiquitous Language", "")
	if len(canvas.ubiquitousLanguage) > 0 {
		*lines = append(*lines, "| Term | Definition |", "| --- | --- |")
		for _, pair := range canvas.ubiquitousLanguage {
			*lines = append(*lines, fmt.Sprintf("| %s | %s |", pair[0], pair[1]))
		}
	} else {
		*lines = append(*lines, "*None*")
	}
	*lines = append(*lines, "")
}

func addListSection(lines *[]string, heading string, items []string) {
	*lines = append(*lines, fmt.Sprintf("## %s", heading), "")
	if len(items) > 0 {
		for _, item := range items {
			*lines = append(*lines, fmt.Sprintf("- %s", item))
		}
	} else {
		*lines = append(*lines, "*None*")
	}
	*lines = append(*lines, "")
}
